pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/NodeInfoPanelController.java

```diff
@@ -146,9 +152,10 @@ private void notifyMetricsAvailable(long numMetrics) {
                                + ((attribute.getValue() != null) ? attribute.getStringValue() : "null"));
         }
 
-        if (node instanceof TypeNode) {
-            result.add("typeIs() = " + ((TypeNode) node).getType());
-        }
+        // TODO maybe put some equivalent to TypeNode inside pmd-core
+        //        if (node instanceof TypeNode) {
+        //            result.add("typeIs() = " + ((TypeNode) node).getType());
```

Comment:
This information was actually extremely useful. How about adding the following getter to java's `AbstractJavaTypeNode`? 

```java
    public String getTypeIs() {
        Class<?> type = getType();
        return type == null ? null : type.getName();
    }
```

Alternatively we could allow the type `Class` as one of the considered return types in `AttributeAxisIterator` - similar to the enum. That would then add a attribute called `type`.

Reply:
Your first workaround could have ramifications... If we expose the attribute, then there would be two ways to query the type of a node, the `typeIs` function and the attribute. The attribute would be strictly less powerful, since it would be a string and we couldn't do e.g. subtyping checks. And having those multiple ways would be confusing to users.

Your second solution would suffer from the exact same drawback if we convert the Class to a string before handing it to the XPath engine. For it to work we should probably hand Class object as a typed value. The attribute would be there, so we could inspect it in the designer and other GUIs. But the value wouldn't be convertible to a string (when executing the query), so that trying to eg compare it to a string (e.g. `@Type = "int"`) wouldn't typecheck. So in effect, the only way to query the type would still be our typeIs and typeIsExactly functions. We could also replace those functions with new functions manipulating the type object directly. E.g.
```ruby
*[@Type = pmd:typeOf("java.util.List")] (: Like typeIsExactly :) 
*[pmd:subsumes(@Type, "java.util.List")] (: Like typeIs :) 
```
But I'm not sure that syntax is clearer.

For now I don't know how much work is required for that to work.

Reply:
Yes, it would be a String-typed attribute. I meant it mostly as a useful information to be displayed in the designer. But bringing this to the designer without adding a XPath attribute will require language neutral support for type systems, something like you put into your TODO comment: a "equivalent of TypeNode in pmd-core"...
I guess that's something for the future then...

Reply:
Note that #1630 brings back the `typeIs` information by checking reflectively if a method named getType is available (see 16267426d407e0037631a2dc87c9818ed065b421). It's a workaround until we decide on a more long-term solution